commit 3ad938e16227692213f76b5b58571cc8549eb171
Author: Ian Bull <irbull>
Date:   Tue Mar 30 22:26:28 2010 +0000

    bug 305237: [repository] Handle operation cancelled exception in execute batch
    https://bugs.eclipse.org/bugs/show_bug.cgi?id=305237

2	0	bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/CompositeArtifactRepository.java
2	0	bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/simple/SimpleArtifactRepository.java
2	0	bundles/org.eclipse.equinox.p2.directorywatcher/src/org/eclipse/equinox/internal/provisional/p2/directorywatcher/CachingArtifactRepository.java
2	0	bundles/org.eclipse.equinox.p2.extensionlocation/src/org/eclipse/equinox/internal/p2/extensionlocation/ExtensionLocationArtifactRepository.java
3	1	bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/p2/repository/artifact/spi/AbstractArtifactRepository.java
2	0	bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/p2/repository/metadata/spi/AbstractMetadataRepository.java
25	0	bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/artifact/repository/BatchExecuteArtifactRepositoryTest.java
2	0	bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/publisher/TestArtifactRepository.java
diff --git a/bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/CompositeArtifactRepository.java b/bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/CompositeArtifactRepository.java
index e0e4b0d..04167f7 100644
--- a/bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/CompositeArtifactRepository.java
+++ b/bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/CompositeArtifactRepository.java
@@ -568,6 +568,8 @@ public class CompositeArtifactRepository extends AbstractArtifactRepository impl
 			try {
 				disableSave = true;
 				runnable.run(monitor);
+			} catch (OperationCanceledException oce) {
+				return new Status(IStatus.CANCEL, Activator.ID, oce.getMessage(), oce);
 			} catch (Throwable e) {
 				result = new Status(IStatus.ERROR, Activator.ID, e.getMessage(), e);
 			} finally {
diff --git a/bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/simple/SimpleArtifactRepository.java b/bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/simple/SimpleArtifactRepository.java
index 1192a5d..1bb37ed 100644
--- a/bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/simple/SimpleArtifactRepository.java
+++ b/bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/simple/SimpleArtifactRepository.java
@@ -1042,6 +1042,8 @@ public class SimpleArtifactRepository extends AbstractArtifactRepository impleme
 			try {
 				disableSave = true;
 				runnable.run(monitor);
+			} catch (OperationCanceledException oce) {
+				return new Status(IStatus.CANCEL, Activator.ID, oce.getMessage(), oce);
 			} catch (Throwable e) {
 				result = new Status(IStatus.ERROR, Activator.ID, e.getMessage(), e);
 			} finally {
diff --git a/bundles/org.eclipse.equinox.p2.directorywatcher/src/org/eclipse/equinox/internal/provisional/p2/directorywatcher/CachingArtifactRepository.java b/bundles/org.eclipse.equinox.p2.directorywatcher/src/org/eclipse/equinox/internal/provisional/p2/directorywatcher/CachingArtifactRepository.java
index 8d3e88a..36bfd20 100644
--- a/bundles/org.eclipse.equinox.p2.directorywatcher/src/org/eclipse/equinox/internal/provisional/p2/directorywatcher/CachingArtifactRepository.java
+++ b/bundles/org.eclipse.equinox.p2.directorywatcher/src/org/eclipse/equinox/internal/provisional/p2/directorywatcher/CachingArtifactRepository.java
@@ -264,6 +264,8 @@ public class CachingArtifactRepository implements IArtifactRepository, IFileArti
 	public IStatus executeBatch(IRunnableWithProgress runnable, IProgressMonitor monitor) {
 		try {
 			runnable.run(monitor);
+		} catch (OperationCanceledException oce) {
+			return new Status(IStatus.CANCEL, Activator.ID, oce.getMessage(), oce);
 		} catch (Exception e) {
 			return new Status(IStatus.ERROR, Activator.ID, e.getMessage(), e);
 		}
diff --git a/bundles/org.eclipse.equinox.p2.extensionlocation/src/org/eclipse/equinox/internal/p2/extensionlocation/ExtensionLocationArtifactRepository.java b/bundles/org.eclipse.equinox.p2.extensionlocation/src/org/eclipse/equinox/internal/p2/extensionlocation/ExtensionLocationArtifactRepository.java
index 8bf8a02..984f563 100644
--- a/bundles/org.eclipse.equinox.p2.extensionlocation/src/org/eclipse/equinox/internal/p2/extensionlocation/ExtensionLocationArtifactRepository.java
+++ b/bundles/org.eclipse.equinox.p2.extensionlocation/src/org/eclipse/equinox/internal/p2/extensionlocation/ExtensionLocationArtifactRepository.java
@@ -252,6 +252,8 @@ public class ExtensionLocationArtifactRepository extends AbstractRepository<IArt
 	public IStatus executeBatch(IRunnableWithProgress runnable, IProgressMonitor monitor) {
 		try {
 			runnable.run(monitor);
+		} catch (OperationCanceledException oce) {
+			return new Status(IStatus.CANCEL, Activator.ID, oce.getMessage(), oce);
 		} catch (Exception e) {
 			return new Status(IStatus.ERROR, Activator.ID, e.getMessage(), e);
 		}
diff --git a/bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/p2/repository/artifact/spi/AbstractArtifactRepository.java b/bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/p2/repository/artifact/spi/AbstractArtifactRepository.java
index 2ae04bd..692c66f 100644
--- a/bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/p2/repository/artifact/spi/AbstractArtifactRepository.java
+++ b/bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/p2/repository/artifact/spi/AbstractArtifactRepository.java
@@ -90,7 +90,7 @@ public abstract class AbstractArtifactRepository extends AbstractRepository<IArt
 	public IArtifactDescriptor createArtifactDescriptor(IArtifactKey key) {
 		return new ArtifactDescriptor(key);
 	}
-	
+
 	public IArtifactKey createArtifactKey(String classifier, String id, Version version) {
 		return new ArtifactKey(classifier, id, version);
 	}
@@ -98,6 +98,8 @@ public abstract class AbstractArtifactRepository extends AbstractRepository<IArt
 	public IStatus executeBatch(IRunnableWithProgress runnable, IProgressMonitor monitor) {
 		try {
 			runnable.run(monitor);
+		} catch (OperationCanceledException oce) {
+			return new Status(IStatus.CANCEL, Activator.ID, oce.getMessage(), oce);
 		} catch (Exception e) {
 			return new Status(IStatus.ERROR, Activator.ID, e.getMessage(), e);
 		}
diff --git a/bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/p2/repository/metadata/spi/AbstractMetadataRepository.java b/bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/p2/repository/metadata/spi/AbstractMetadataRepository.java
index 2dfbca5..b526bb8 100644
--- a/bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/p2/repository/metadata/spi/AbstractMetadataRepository.java
+++ b/bundles/org.eclipse.equinox.p2.repository/src/org/eclipse/equinox/p2/repository/metadata/spi/AbstractMetadataRepository.java
@@ -143,6 +143,8 @@ public abstract class AbstractMetadataRepository extends AbstractRepository<IIns
 	public IStatus executeBatch(IRunnableWithProgress runnable, IProgressMonitor monitor) {
 		try {
 			runnable.run(monitor);
+		} catch (OperationCanceledException oce) {
+			return new Status(IStatus.CANCEL, Activator.ID, oce.getMessage(), oce);
 		} catch (Exception e) {
 			return new Status(IStatus.ERROR, Activator.ID, e.getMessage(), e);
 		}
diff --git a/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/artifact/repository/BatchExecuteArtifactRepositoryTest.java b/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/artifact/repository/BatchExecuteArtifactRepositoryTest.java
index ba65a7b..ca1065c 100644
--- a/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/artifact/repository/BatchExecuteArtifactRepositoryTest.java
+++ b/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/artifact/repository/BatchExecuteArtifactRepositoryTest.java
@@ -302,6 +302,31 @@ public class BatchExecuteArtifactRepositoryTest extends AbstractProvisioningTest
 	/*
 	 * This tests that exceptions are properly propagated for a CompositeRepository
 	 */
+	public void testBatchProcessingCancelled() {
+		try {
+			Map properties = new HashMap();
+			repositoryFile = getTempFolder();
+			repositoryURI = repositoryFile.toURI();
+			SimpleArtifactRepository repo = (SimpleArtifactRepository) getArtifactRepositoryManager().createRepository(repositoryURI, "My Repo", IArtifactRepositoryManager.TYPE_SIMPLE_REPOSITORY, properties);
+			IProgressMonitor monitor = new NullProgressMonitor();
+			monitor.setCanceled(true);
+
+			IStatus status = repo.executeBatch(new IRunnableWithProgress() {
+				public void run(IProgressMonitor monitor) {
+					if (monitor.isCanceled())
+						throw new OperationCanceledException();
+				}
+			}, monitor);
+
+			assertTrue(status.getSeverity() == IStatus.CANCEL);
+		} catch (Exception e) {
+			fail("Test failed", e);
+		}
+	}
+
+	/*
+	 * This tests that exceptions are properly propagated for a CompositeRepository
+	 */
 	public void testBatchProcessingExceptionsComposite() {
 		try {
 			FailingCompositeArtifactRepository compositeArtifactRepository = new FailingCompositeArtifactRepository(getArtifactRepositoryManager(), "foo", new URI("http://foo.bar"), null);
diff --git a/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/publisher/TestArtifactRepository.java b/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/publisher/TestArtifactRepository.java
index 8550992..7b1ae7b 100644
--- a/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/publisher/TestArtifactRepository.java
+++ b/bundles/org.eclipse.equinox.p2.tests/src/org/eclipse/equinox/p2/tests/publisher/TestArtifactRepository.java
@@ -329,6 +329,8 @@ public class TestArtifactRepository implements IArtifactRepository {
 	public IStatus executeBatch(IRunnableWithProgress runnable, IProgressMonitor monitor) {
 		try {
 			runnable.run(monitor);
+		} catch (OperationCanceledException oce) {
+			return new Status(IStatus.CANCEL, "org.eclipse.equinox.p2.tests.publisher", oce.getMessage(), oce);
 		} catch (Exception e) {
 			return new Status(IStatus.ERROR, "org.eclipse.equinox.p2.tests.publisher", e.getMessage(), e);
 		}
